components/engine_traits/src/lib.rs

```diff
@@ -1,79 +1,108 @@
 
-mod errors;
-pub use crate::errors::*;
-mod peekable;
-pub use crate::peekable::*;
-mod iterable;
-pub use crate::iterable::*;
-mod mutable;
-pub use crate::mutable::*;
-mod cf;
-pub use crate::cf::*;
-mod engine;
-pub use crate::engine::*;
-mod options;
-pub use crate::options::*;
-mod db_options;
-mod util;
-pub use crate::db_options::*;
+// These modules contain traits that need to be implemented by engines, either
+// they are required by KvEngine or are an associated type of KvEngine. It is
+// recommended that engines follow the same module layout.
+//
+// Many of these define "extension" traits, that end in `Ext`.
+
 mod cf_handle;
 pub use crate::cf_handle::*;
 mod cf_options;
 pub use crate::cf_options::*;
+mod db_options;
+pub use crate::db_options::*;
+mod engine;
+pub use crate::engine::*;
 mod import;
 pub use import::*;
+mod snapshot;
+pub use crate::snapshot::*;
+mod write_batch;
+pub use crate::write_batch::*;
 
-pub const DATA_KEY_PREFIX_LEN: usize = 1;
-
-// In our tests, we found that if the batch size is too large, running delete_all_in_range will
-// reduce OLTP QPS by 30% ~ 60%. We found that 32K is a proper choice.
-pub const MAX_DELETE_BATCH_SIZE: usize = 32 * 1024;
-
-#[derive(Clone, Debug)]
-pub struct KvEngines<K, R> {
-    pub kv: K,
-    pub raft: R,
-    pub shared_block_cache: bool,
-}
-
-impl<K: KvEngine, R: KvEngine> KvEngines<K, R> {
-    pub fn new(kv_engine: K, raft_engine: R, shared_block_cache: bool) -> Self {
-        KvEngines {
-            kv: kv_engine,
-            raft: raft_engine,
-            shared_block_cache,
-        }
-    }
+// These modules contain more general traits, some of which may be implemented
+// by multiple types.
 
-    pub fn write_kv(&self, wb: &K::Batch) -> Result<()> {
-        self.kv.write(wb)
-    }
-
-    pub fn write_kv_opt(&self, wb: &K::Batch, opts: &WriteOptions) -> Result<()> {
-        self.kv.write_opt(opts, wb)
-    }
+mod iterable;
+pub use crate::iterable::*;
+mod mutable;
+pub use crate::mutable::*;
+mod peekable;
+pub use crate::peekable::*;
 
-    pub fn sync_kv(&self) -> Result<()> {
-        self.kv.sync()
-    }
+// These modules contain support code that does not need to be implemented by
+// engines.
 
-    pub fn write_raft(&self, wb: &R::Batch) -> Result<()> {
-        self.raft.write(wb)
-    }
+mod cfdefs;
+pub use crate::cfdefs::*;
+mod engines;
+pub use engines::*;
+mod errors;
+pub use crate::errors::*;
+mod options;
+pub use crate::options::*;
+pub mod util;
 
-    pub fn write_raft_opt(&self, wb: &R::Batch, opts: &WriteOptions) -> Result<()> {
-        self.raft.write_opt(opts, wb)
-    }
+pub const DATA_KEY_PREFIX_LEN: usize = 1;
 
-    pub fn sync_raft(&self) -> Result<()> {
-        self.raft.sync()
-    }
-}
+// In our tests, we found that if the batch size is too large, running delete_all_in_range will
+// reduce OLTP QPS by 30% ~ 60%. We found that 32K is a proper choice.
+pub const MAX_DELETE_BATCH_SIZE: usize = 32 * 1024;
```

Comment:
I think this is a RocksDB relative variable, better to move to engine_rocks.

Reply:
The const is duplicated from the engine crate and appears to be unused so I've just deleted it.